src/pipelines/worldpop/worldpop_population.py

```diff
@@ -0,0 +1,101 @@
+from pandas import DataFrame, Series, read_csv, isnull
+
+from lib.concurrent import thread_map
+from lib.net import download
+from lib.pipeline import DataSource
+from lib.cast import age_group
+
+
+class WorldPopPopulationDataSource(DataSource):
+    """ Retrieves the requested properties from Wikidata for all items in metadata.csv """
+
+    url_strat_tpl = (
+        "ftp://ftp.worldpop.org.uk/GIS"
+        "/AgeSex_structures/Global_2000_2020/2020/"
+        "{code}/{code_lower}_{sex}_{age_bucket}_2020.tif"
+    )
+
+    url_total_tpl = (
+        "ftp://ftp.worldpop.org.uk/GIS"
+        "/Population/Global_2000_2020/2020/{code}/{code_lower}_ppp_2020.tif"
+    )
+
+    @staticmethod
+    def aggregate_population(data: Series) -> Dict[str, int]:
+        age_bucket_size = 10
+        age_bucket_count = 10
+        age_bucket_pairs = [
+            (i * age_bucket_size, (i + 1) * age_bucket_size - 1) for i in range(age_bucket_count)
+        ]
+
+        agg = {
+            "key": data["key"],
+            "population": 0,
+            "population_male": 0,
+            "population_female": 0,
+            **{f"population_age_{lo:02d}_{hi:02d}": 0 for lo, hi in age_bucket_pairs},
+        }
+        for col, val in data.iteritems():
+            # Skip over the key item
+            if col == "key":
+                continue
+
+            # Total population is the sum of all populations
+            agg["population"] += val
+
+            # Get age/sex info from column name
+            sex, age = col.split("_", 2)
+
+            # Add male and female populations separately
+            if sex == "m":
+                agg["population_male"] += val
+            if sex == "f":
+                agg["population_female"] += val
+
+            # Go over all the age buckets and add them separately
+            age_bucket = age_group(int(age)).replace("-", "_")
+            agg[f"population_age_{age_bucket}"] = val
+
+        return agg
+
+    def parse(self, sources: List[str], aux: Dict[str, DataFrame], **parse_opts) -> DataFrame:
+
+        # WorldPop is pre-computed for all regions using Earth Engine, a Google product:
+        # https://code.earthengine.google.com/?scriptPath=users%2Fowahltinez%2Fopen-covid-admin%3Aprocess
+
+        # Compute the estimated total male population
+        records = aux["worldpop"].apply(self.aggregate_population, axis=1)
+        data = DataFrame.from_records(records).merge(aux["worldpop"])
+
+        # Keep only records which are part of the index
+        data = data.merge(aux["metadata"][["key"]])
+
+        # WorldPop only provides data for people up to 80 years old, but we want 10-year buckets
+        # until 90, and 90+ instead. We estimate that, within the 80+ group, 80% are 80-90 and
+        # 20% are 90+. This is based on multiple reports, for example:
+        # https://www.statista.com/statistics/281174/uk-population-by-age
+        data["population_age_90_99"] = data["population_age_80_89"] * 0.2
```

Comment:
I'm a bit skeptical of this approach. While I don't doubt this claim is valid, splitting this bucket may introduce statistical noise. I think it would be more appropriate to document that the last bucket is age 80+ and allow clients using this data to decide how they want to handle that rather than making the decision for them. If this splitting is done to establish consistency with other data sources it may be worth relaxing that restriction, or normalizing these buckets.

Also, is this data for the UK only?

Reply:
Good point Anthony.

Do we have overall population 80-89 and 90+ data split for any sources?  If we only have this split for other measures and not for overall population counts, especially if it's for a small fraction of sources, I'd prefer to change the upper cutoff from 90+ to 80+ across the board to be able to simply and consistently present data without requiring this imputation, and guard against problems caused by combining actual splits of 80-89/90+ for particular measures with the imputed 80%/20% split for overall population.

If we do think there's significant value in maintaining the 80-89/90+ split, we should clearly document the imputation in a way that's obvious to anyone who pulls or downloads the file.

Anthony, what do you mean by "normalizing these buckets"?

Long term, as we move to Cloud (and perhaps to a relational database?), I hope we can provide a flexible way for users to produce a file like demographics.csv or master.csv for their chosen set of regions and fields on the fly from the raw data sources, either restricted to the regions that already have consistency in age bucketing, or making the explicit choice to impute / normalize data.

Reply:
You are totally right about this, and it's pretty hand-wavy. The intent is to provide two versions: one with the data as close to the source of truth as possible (`worldpop.csv`) and another with "useful transformations" applied to it (in this case, `demographics.csv`). I believe that providing a consistent set of buckets across all sources which are stratified can be extremely useful, and worth the hand-waviness.

Reply:
I agree with the general idea and with providing a consistent set of buckets across all sources, but would probably (subject to the answer to questions in my earlier comment) prefer to set the highest bucket in that set to 80+ to avoid having to impute the 80%/20% split here.

Reply:
Discussed offline